src/daft-core/src/array/ops/utf8.rs

```diff
@@ -1383,6 +1384,49 @@ impl Utf8Array {
         ))
     }
 
+    // Uses the Aho-Corasick algorithm to count occurrences of a number of patterns.
+    pub fn count_matches(
+        &self,
+        patterns: &Self,
+        whole_word: bool,
+        case_sensitive: bool,
```

Comment:
I wonder if it'd be simpler to instead accept a regex?

the [regex](https://docs.rs/regex/1.10.5/regex/) crate uses `aho-corasick` under the hood, so the perf implications should be negligible as long as we're only compiling the regex once. 

Regex would likely be a lot more intuitive from an end user perspective as well. 

```py
res = s.str.count_matches('(fox|over|lazy dog|dog)').to_pylist()
```

Reply:
Yeah, I'll change it to that. I didn't originally do this because I was worried about performance - I'll run some tests to make sure it isn't too affected.

Reply:
Unfortunately it does seem like using a regex is around 7x slower. I guess it just can't handle the large pattern created by concatenating the list of strings. So I think I'll keep it this way for now.

Reply:
Even though it's slower, I think we should still allow for regex just for usability sake. 

We can do this in a follow up PR, and have the python frontend map to a different backend implementation


`count_matches('<text>') -> count_matches`
`count_matches(r'<pattern>') -> count_matches_regex`